src/Build/Xml/ProjectXmlUtilities.XmlElementChildIterator.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Xml;
+using Microsoft.Build.Construction;
+
+namespace Microsoft.Build.Internal
+{
+    partial class ProjectXmlUtilities
+    {
+        // Iterating an element's nodes allocates an non-trivial amount of data in certain
+        // large solutions with lots of targets, so we have our own struct-based iterator
+        // that avoids unneeded GC pressure.
+        //
+        // Deliberately not implement IEnumerable/IEnumerator to avoid accidental boxing
+        internal struct XmlElementChildIterator
+        {
+            private readonly XmlElementWithLocation _element;
+            private readonly bool _throwForInvalidNodeTypes;
+            private XmlElementWithLocation _current;
+            private bool _isFirst;
+
+            internal XmlElementChildIterator(XmlElementWithLocation element, bool throwForInvalidNodeTypes)
+            {
+                _element = element;
+                _throwForInvalidNodeTypes = throwForInvalidNodeTypes;
+                _current = null;
+                _isFirst = true;
+            }
+
+            public bool MoveNext()
+            {
+                if (_isFirst)
+                {
+                    _isFirst = false;
+                    _current = GetNextNode(_element.FirstChild);
+                }
+                else if (_current != null)
+                {
+                    _current = GetNextNode(_current.NextSibling);
+                }
+
+                return _current != null;
+            }
+
+            public XmlElementChildIterator GetEnumerator()
+            {
+                return this;
```

Comment:
This may be a standard pattern, but is it correct to have the enumerable and the enumerator be the same object?  It seems like the contract would be that you can call GetEnumerator() multiple times on the same enumerable and you should get different enumerators / iterators with different state.

Reply:
It's a pattern used by LINQ- http://index/?query=Enumerable.Select&rightProject=System.Core&file=System%5CLinq%5CEnumerable.cs&line=90. I was a little worried about increasing the size to handle this case (because it will never be hit), but I noticed they use a "state" field to track this and I might do the same thing.